core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainerFactory.scala

```diff
@@ -60,10 +62,14 @@ class DockerContainerFactory(instance: InvokerInstanceId,
                                userProvidedImage: Boolean,
                                memory: ByteSize,
                                cpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+    val registryConfig =
+      ContainerFactory.resolveRegisterConfig(userProvidedImage, runtimesRegistryConfig, userImagesRegistryConfig)
+    val image =
+      if (userProvidedImage && registryConfig.url.isEmpty) Left(actionImage)
+      else Right(actionImage.localImageName(registryConfig.url))
```

Comment:
I have the expression that this code can be streamlined as follows to simplify the condition in line 68 (`(userProvidedImage && registryConfig.url.isEmpty)`):

```
   val registryConfig =
      ContainerFactory.resolveRegisterConfig(userProvidedImage, runtimesRegistryConfig, userImagesRegistryConfig)
    val imageRef = actionImage.localImageName(registryConfig.url)
    val image = if (userProvidedImage) Left(imageRef) else Right(imageRef)
```

My understanding is that `actionImage.localImageName()` would just leave out the registry part if no registry urls are specified.

Reply:
Would this lead to the same result - as done in the Kube container factory below?

```
   val registryConfig =
      ContainerFactory.resolveRegisterConfig(userProvidedImage, runtimesRegistryConfig, userImagesRegistryConfig)
    val imageRef = ContainerFactory.resolveImage(actionImage, registryConfig)
    val image = if (userProvidedImage) Left(imageRef) else Right(imageRef)
```

Reply:
I guess you meant it to be:
```
val registryConfig =
      ContainerFactory.resolveRegisterConfig(userProvidedImage, runtimesRegistryConfig, userImagesRegistryConfig)
val imageRef = actionImage.localImageName(registryConfig.url)
val image = if (userProvidedImage) Left(actionImage) else Right(imageRef)
```
(`actionImage` in Left, not `imageRef`)

However the condition is not the same as before anymore. In the case of `userProvidedImage==true`, the raw `imageName` object is passed into DockerContainer, which will pull the "public image name" (without custom blackbox registry), see [DockerContainer.scala#L105](https://github.com/apache/openwhisk/blob/master/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala#L105). Hence I added a condition for registryConfig being empty or not.

In general this flow does not fit well into the existing flow of DockerContainerFactory. I am working on an enhancement to get rid of `publicImageName`, make `localImageName` (to be renamed) a single image name resolver, and apply it throughout the code base. I will push that change for review.

Reply:
My understanding is that the following behaviour is desired for blackbox (i.e. user provided) images identified by a user provided image reference (see https://github.com/docker/distribution/blob/master/reference/reference.go):

* If the image reference contains a domain, keep the image reference as is to pull from the repository in the image reference.
* If the image reference contains NO domain and NO global user image registry is configured (`registryConfig.url.isEmpty == true`), keep the image reference as is. It's up to the used container factory to convert to a canonical reference including a domain and path.
* If the image reference contains NO domain and a global user image registry is configured (`registryConfig.url.isEmpty == false`), add the global user image registry as domain to the image reference.

I think this is exactly the behaviour that following code implements:

```
   val registryConfig =
      ContainerFactory.resolveRegisterConfig(userProvidedImage, runtimesRegistryConfig, userImagesRegistryConfig)
    val imageRef = ContainerFactory.resolveImage(actionImage, registryConfig)
```

That's why I proposed this code because it already addresses the three cases listed above.

Reply:
alright, now I got what you meant, thanks for explaining.
My intention with "If the image reference contains a domain", is that if there is a global user image registry, it would overwrite the image domain. I thought of this in a multi-cloud / multi-region setup in which a cluster for a specific cloud / region could force to download *all* images from its desired registry.
But I also agree with you that keeping the domain in image reference (if existed) no matter what global registry is, would be more trivial to users.

Reply:
Yes, since your PR does not provide a list with mappings for registry domains but only a single user image repository domain, all user provided image refs would be overwritten - regardless whether the user provided image ref already contains a domain or not.

This current approach would lead to:

With a user registry config like `uk.icr.io`, this would lead to situations like:

* User registry config empty, user-provided image ref `image:latest` => Docker container factory pulls `docker.io/library/image:latest`.
* User registry config empty, user-provided image ref `docker.io/library/image:latest` => Docker container factory pulls `docker.io/library/image:latest`.
* User registry config `uk.icr.io`, user-provided image ref `image:latest` => Docker container factory pulls `uk.icr.io/image:latest`.
  The user would probably expect to pull `uk.icr.io/library/image:latest` instead - but `docker pull` only inserts the "official" repository path `library` if `docker.io` is specified as domain. See https://github.com/docker/distribution/blob/14b96e55d84c367ebab6caf9f0086de0876eec72/reference/normalize.go#L88-L105.
* User registry config `uk.icr.io`, user-provided image ref `path/image:latest` => Docker container factory pulls `uk.icr.io/path/image:latest`.
* User registry config `uk.icr.io`, user-provided image ref `docker.io/path/image:latest` => Docker container factory pulls `uk.icr.io/path/image:latest`.
  The user would certainly not expect this outcome. If you provide an image ref with a domain, you would not expect that it is replaced.

Reply:
ok, I think what you said makes sense. thanks @sven-lange-last.
Please see the latest PR version.